examples/errors.rs

```diff
@@ -34,49 +30,53 @@ impl Display for Error {
-        warp::path("nope").and_then(|| Err::<StatusCode, _>(warp::reject::custom(Error::Nope)));
+        warp::path("nope").and_then(|| async { Err::<StatusCode, _>(warp::reject::custom(Error::Nope))});
 
     let routes = warp::get2()
         .and(hello.or(oops).or(nope))
         .recover(customize_error);
 
-    warp::serve(routes).run(([127, 0, 0, 1], 3030));
+    warp::serve(routes).run(([127, 0, 0, 1], 3030)).await;
 }
 
 // This function receives a `Rejection` and tries to return a custom
 // value, othewise simply passes the rejection along.
-fn customize_error(err: Rejection) -> Result<impl Reply, Rejection> {
-    if let Some(&err) = err.find_cause::<Error>() {
-        let code = match err {
-            Error::Nope => StatusCode::BAD_REQUEST,
-            Error::Oops => StatusCode::INTERNAL_SERVER_ERROR,
-        };
-        let msg = err.to_string();
+fn customize_error(err: Rejection) -> impl Future< Output = Result<impl Reply, Rejection>> {
+    let err = {
+        if let Some(&err) = err.find_cause::<Error>() {
+            let code = match err {
+                Error::Nope => StatusCode::BAD_REQUEST,
+                Error::Oops => StatusCode::INTERNAL_SERVER_ERROR,
+            };
+            let msg = err.to_string();
 
-        let json = warp::reply::json(&ErrorMessage {
-            code: code.as_u16(),
-            message: msg,
-        });
-        Ok(warp::reply::with_status(json, code))
-    } else if let Some(_) = err.find_cause::<warp::reject::MethodNotAllowed>() {
-        // We can handle a specific error, here METHOD_NOT_ALLOWED,
-        // and render it however we want
-        let code = StatusCode::METHOD_NOT_ALLOWED;
-        let json = warp::reply::json(&ErrorMessage {
-            code: code.as_u16(),
-            message: "oops, you aren't allowed to use this method.".into(),
-        });
-        Ok(warp::reply::with_status(json, code))
-    } else {
-        // Could be a NOT_FOUND, or any other internal error... here we just
-        // let warp use its default rendering.
-        Err(err)
-    }
+            let json = warp::reply::json(&ErrorMessage {
+                code: code.as_u16(),
+                message: msg,
+            });
+            Ok(warp::reply::with_status(json, code))
+        } else if let Some(_) = err.find_cause::<warp::reject::MethodNotAllowed>() {
+            // We can handle a specific error, here METHOD_NOT_ALLOWED,
+            // and render it however we want
+            let code = StatusCode::METHOD_NOT_ALLOWED;
+            let json = warp::reply::json(&ErrorMessage {
+                code: code.as_u16(),
+                message: "oops, you aren't allowed to use this method.".into(),
+            });
+            Ok(warp::reply::with_status(json, code))
+        } else {
+            // Could be a NOT_FOUND, or any other internal error... here we just
+            // let warp use its default rendering.
+            Err(err)
+        }
+    };
+    async { err }
```

Comment:
Same as above, use `futures::future::ready` so we're not paying the generator cost.

Reply:
thanks! had missed that